File: java/testsuite/OfflineExamples/RaytracingOfflineExample/src/RaytracerExecutor.java
import vsdk.toolkit.environment.material.RendererConfiguration;
import vsdk.toolkit.environment.scene.SimpleSceneSnapshot;
import vsdk.toolkit.gui.feedback.ProgressMonitor;
import vsdk.toolkit.media.RGBImageUncompressed;
import vsdk.toolkit.render.SimpleRaytracer;

interface RaytracerExecutor {
    void run(SimpleRaytracer visualizationEngine,
             RGBImageUncompressed resultingImage,
             RendererConfiguration rendererConfiguration,
             SimpleSceneSnapshot sceneSnapshot,
             ProgressMonitor reporter);
}

File: typescript/testsuite/OfflineExamples/RaytracingOfflineExample/src/RaytracerExecutor.ts
import type {
    ProgressMonitor,
    RGBImageUncompressed,
    RendererConfiguration,
    SimpleRaytracer,
    SimpleSceneSnapshot,
} from "@vitral/base";

/**
Java declares `run` as `void`. The parallel implementation needs worker
threads, which Node only exposes asynchronously, so `run` may also answer a
promise here and `RaytracerSimple` awaits it. The serial implementation still
returns nothing, exactly as in Java.
*/
export interface RaytracerExecutor {
    run(
        visualizationEngine: SimpleRaytracer,
        resultingImage: RGBImageUncompressed,
        rendererConfiguration: RendererConfiguration,
        sceneSnapshot: SimpleSceneSnapshot,
        reporter: ProgressMonitor,
    ): void | Promise<void>;
}
